### PR TITLE
[ENG-2123][ENG-2268] Legacy indicators for storage usage

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Let a user know when they are at or approaching a their storage limit.

## Changes

- make sure we are consistently using GB vs. GiB (detour, but since current storage usage is in GBs this really has to be too for ease of formatting)
- pass storage status info into contextvar in the mako template and format
- pass storage status info into fangorn
- throw storage usage error growl on upload 409 response

Screen Shot 2020-09-10 at 2 15 08 PM
Screen Shot 2020-09-10 at 2 15 03 PM
Screen Shot 2020-09-10 at 2 14 13 PM
Screen Shot 2020-09-10 at 2 13 53 PM

## QA Notes

Note: Since https://openscience.atlassian.net/browse/ENG-2122 is still in progress the growlbox that should appear when a user uploads a file over the limit shouldn't actually be trigger-able. 

  - Does this change require a data migration? If so, what data will we migrate?
No
  - What is the level of risk?
Low
    - Any permissions code touched?
No
    - Is this an additive or subtractive change, other?
additive
  - How can QA verify? (Through UI, API, AdminApp or AdminAdminApp?)
UI
  - What features or workflows might this change impact?
None?
  - How will this impact performance?
It shouldn't noticably

Pages affected:

** Project Page **

Screen Shot 2020-09-18 at 1 28 34 PM

** Files  Page **

Screen Shot 2020-09-18 at 11 12 50 AM

and check all different places fangorn is used to ensure it's appropriate.

## Documentation

No new docs, I think.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-2123
https://openscience.atlassian.net/browse/ENG-2268

### CODE REVIEW

File: website/static/js/fangorn.js [hunk 1]
```diff
@@ -1935,18 +1935,33 @@ var FGItemButtons = {
...
                     m.component(FGButton, {
                         onclick: function () {
                             mode(toolbarModes.ADDFOLDER);
                         },
                         icon: 'fa fa-plus',
                         className: 'text-success'
-                    }, 'Create Folder'));
+                    }, 'Create Folder')
+                );
+                if(window.contextVars.node.storageLimitsStatus) {
```
Comment (brianjgeiger) [inline website/static/js/fangorn.js:1946]: Drive-by: Did you disable the dropzone for OSF Storage when the limit was exceeded?

Comment (Johnetordoff) [reply website/static/js/fangorn.js:1946]: It doesn't have a visual indicator that the dropzone is disabled, but if you try to drop it produces a growlbox with message, (or it will once WB returns a 409 as the response).

Comment (brianjgeiger) [reply website/static/js/fangorn.js:1946]: If possible, it would be better to disable the dropzone, to keep people from attempting to upload files, which will save bandwidth costs, and keep the users happier by not wasting their time. Don't go through heroics for this, but it would be better.

File: website/project/utils.py [hunk 2]
```diff
@@ -118,3 +118,43 @@ def sizeof_fmt(num, suffix='B'):
...
+    elif status is settings.StorageLimits.APPROACHING_PUBLIC:
+        return {
+            'class': 'btn-warning  storage-warning',
+            'text': f'This node is approaching its limit of {over_public_size}.'
+        }
+    elif status is settings.StorageLimits.OVER_PUBLIC:
+        return {
+            'class': 'btn-danger  storage-warning',
+            'text': f'This node has exceeded its limit of {over_public_size}.'
+        }
```
Comment (brianjgeiger) [inline website/project/utils.py:156]: Here is the [updated language](https://docs.google.com/document/d/13dZBXO_JqZOcf4fe-HKCk_YDMA5PYokfOOw4KB3jbEY/edit?ts=5f47d7b8#bookmark=id.ga5sxbogg83u) for these and the error notification. Check with Kayla to see if she has a link to the help docs that we can use.

File: website/static/js/fangorn.js (hunk 1 above)
Comment (Johnetordoff) [reply website/static/js/fangorn.js]: Yeah I tried a few different things, but it's too difficult to do this dynamically so that uploading will work once something is deleted and storage usage goes down.

File: website/static/js/fangorn.js [hunk 3]
```diff
@@ -1010,8 +1016,8 @@ function _fangornDropzoneError(treebeard, file, message, xhr) {
 
     if (file.isDirectory) {
         msgText = 'Cannot upload folders.';
-    } else if (xhr && xhr.status === 507) {
```
Comment (Johnetordoff) [inline website/static/js/fangorn.js:1013]: This can't actually be hit

File: osf_tests/test_storage_usage_limits.py [hunk 4]
```diff
@@ -22,7 +22,7 @@ def test_limit_default(self, node):
         assert node.storage_limit_status == StorageLimits.DEFAULT
 
     def test_storage_limits(self, node):
-        GBs = 1024 ** 3.0
+        GBs = 10 ** 9
```
Comment (brianjgeiger) [inline osf_tests/test_storage_usage_limits.py:25]: You are using GBs in some places and GiBs in others. Maybe we should put the GiB into `website.settings.defaults` and just import that into wherever you need to use it, so we have one place it's defined.

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
Looking pretty good. You need to clean up inconsistencies with GB vs GiB, there's a growl box that needs a real message, and possibly a change in an error code.

File: website/static/js/fangorn.js [hunk 5]
```diff
@@ -1010,8 +1016,8 @@ function _fangornDropzoneError(treebeard, file, message, xhr) {
...
-        msgText = 'Cannot upload file due to insufficient storage.';
+    } else if (xhr && xhr.status === 409) {
```
[the remark this answers is not in the document]
Comment (UdayVarkhedkar) [reply website/static/js/fangorn.js:1019]: Sounds good to me!

File: website/settings/defaults.py [hunk 6]
```diff
@@ -1967,6 +1967,8 @@ class CeleryConfig:
 STORAGE_LIMIT_PUBLIC = 50
 STORAGE_LIMIT_PRIVATE = 5
 
+GBs = 10 ** 9
+GiBs = 1024 ** 3
```
Comment (brianjgeiger) [inline website/settings/defaults.py:1971]: Why are we using GBs in some places and GiBs in others?

File: website/project/utils.py [hunk 7]
```diff
@@ -118,3 +118,41 @@ def sizeof_fmt(num, suffix='B'):
...
+    public_limit = node.custom_storage_usage_limit_public or settings.STORAGE_LIMIT_PUBLIC
+    private_limit = node.custom_storage_usage_limit_private or settings.STORAGE_LIMIT_PRIVATE
+
+    over_public_size = sizeof_fmt(int(public_limit * settings.GBs))
+    over_private_size = sizeof_fmt(int(private_limit * settings.GBs))
+
+    if status is settings.StorageLimits.APPROACHING_PRIVATE:
+        return {
+            'class': 'btn-warning storage-warning',
+            'text': f'This project is approaching the storage limit of {over_private_size}. To learn more about limits and alternative storage options visit https://help.osf.io/.',
```
Comment (brianjgeiger) [inline website/project/utils.py:134]: According to the [requirements](https://docs.google.com/document/d/13dZBXO_JqZOcf4fe-HKCk_YDMA5PYokfOOw4KB3jbEY/edit?ts=5f47d7b8#bookmark=id.ga5sxbogg83u), this should be "This project/component is approaching the storage limit for OSF Storage. To learn more about limits and alternative storage options visit our help docs."  And “visit our help docs” is a stable link to a help doc with options and instructions. You should be able to get that link from Kayla. The rest of these should be similarly updated.

Or can we not do a link in the help text?

File: website/static/js/fangorn.js [hunk 8]
```diff
@@ -1011,7 +1015,7 @@ function _fangornDropzoneError(treebeard, file, message, xhr) {
     if (file.isDirectory) {
         msgText = 'Cannot upload folders.';
     } else if (xhr && xhr.status === 507) {
-        msgText = 'Cannot upload file due to insufficient storage.';
+        msgText = 'This file cannot be added to this project because it would exceed the storage limit for OSF Storage';
```
Comment (brianjgeiger) [inline website/static/js/fangorn.js:1018]: "This file cannot be added to this project/component because it would exceed the storage limit for OSF Storage"

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
A question about consistency, and double-checking what was in the requirements for help text.

File: website/project/utils.py (hunk 7 above)
Comment (Johnetordoff) [reply website/project/utils.py:134]: We can't do a link like that in tooltip text, it's not clickable.

File: website/settings/defaults.py (hunk 6 above)
Comment (Johnetordoff) [reply website/settings/defaults.py:1971]: It depends on the context, some places GiBs were being used and just labeled GBs I went through the repo making sure the distinction was explicit.

Comment (brianjgeiger) [reply website/settings/defaults.py:1971]: But _shouldn't_ we be using the same measure in different places?

Comment (Johnetordoff) [reply website/settings/defaults.py:1971]: Well I can't with any certainty what should used in `scripts/osfstorage/usage_audit.py` or `osf/management/commands/export_user_account.py.` I just didn't want to change the behavior. The storage usage cap must be is GBs, because the `storage_usage` field is stored in GBs, because our size info is in GBs, but if you are just running a management command or script, I have to assume the behavior should just be maintained regardless of unit, I'm just making it explicit.

Comment (brianjgeiger) [reply website/settings/defaults.py:1971]: If all the new things are GiBs, then could we just settle on GBs throughout the codebase?

Comment (Johnetordoff) [reply website/settings/defaults.py]: Well i just ran it through travis with no test fails so LGTM.

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
Where is the help text for being over the storage limit? I only see help text for approaching the storage limit. "This project/component exceeds the storage limit for OSF Storage. To learn more about limits and alternative storage options visit our help docs."

**Review by brianjgeiger — APPROVED** [review]

**Review by corbinSanders — APPROVED** [review]
LGTM! ⛵

File: website/templates/project/project.mako [hunk 9]
```diff
@@ -33,13 +33,17 @@
...
             </div>
             <div class="clearfix visible-md-block"></div>
-            <div class="col-lg-4">
+            <div class="col-lg-6">
                 <div class="btn-toolbar node-control pull-right">
-                    % if node.get('storage_usage'):
                     <div class="btn-group">
-                        <button style="pointer-events: auto;" class="btn disabled" data-toggle="tooltip" data-placement="bottom" title="This is the amount of OSF Storage used for this project, including all versions of files.">${node['storage_usage']}</button>
+                        % if node.get('storage_limit_status') and node['storage_limit_status']['text']:
+                            <a href=" https://help.osf.io/hc/en-us/articles/360054528874-OSF-Storage-Caps"  target="_blank" class="btn ${node['storage_limit_status']['class']}"  data-toggle="tooltip" data-placement="bottom" title="This project/component is ${node['storage_limit_status']['text']} the storage limit for OSF Storage. To learn more about limits and alternative storage options click on this icon."><i class="fa fa-exclamation-triangle"></i></a>
```
Comment (brianjgeiger) [inline website/templates/project/project.mako:40]: Drive-by

```suggestion
                            <a href="https://help.osf.io/hc/en-us/articles/360054528874-OSF-Storage-Caps"  target="_blank" class="btn ${node['storage_limit_status']['class']}"  data-toggle="tooltip" data-placement="bottom" title="This project/component is ${node['storage_limit_status']['text']} the storage limit for OSF Storage. To learn more about limits and alternative storage options click on this icon."><i class="fa fa-exclamation-triangle"></i></a>
```

**Review by UdayVarkhedkar — APPROVED** [review]
LGTM

**Comment by brianjgeiger** [comment]
Merged with feature/storage-caps
